spring-data-jpa/src/test/java/org/springframework/data/jpa/repository/config/JpaRepositoryRegistrationAotProcessorUnitTests.java

```diff
@@ -132,6 +134,104 @@ public List<String> getManagedPackages() {
 		assertThat(contributor.getMetamodel().managedType(Person.class)).isNotNull();
 	}
 
+	@Test // GH-3899
+	void repositoryProcessorShouldEnableAotRepositoriesByDefaultWhenAotIsEnabled() {
+
+		GenerationContext ctx = new DefaultGenerationContext(new ClassNameGenerator(ClassName.OBJECT),
+				new InMemoryGeneratedFiles());
+
+		GenericApplicationContext context = new GenericApplicationContext();
+		
+		System.setProperty(AotDetector.AOT_ENABLED, "true");
+		
+		try {
+			JpaRepositoryContributor contributor = new JpaRepositoryConfigExtension.JpaRepositoryRegistrationAotProcessor()
+					.contribute(new DummyAotRepositoryContext(context) {
+						@Override
+						public Set<Class<?>> getResolvedTypes() {
+							return Collections.singleton(Person.class);
+						}
+					}, ctx);
+
+			assertThat(contributor).isNotNull();
+		} finally {
+			System.clearProperty(AotDetector.AOT_ENABLED);
+		}
+	}
+
+	@Test // GH-3899
+	void repositoryProcessorShouldNotEnableAotRepositoriesByDefaultWhenAotIsDisabled() {
+
+		GenerationContext ctx = new DefaultGenerationContext(new ClassNameGenerator(ClassName.OBJECT),
+				new InMemoryGeneratedFiles());
+
+		GenericApplicationContext context = new GenericApplicationContext();
+
+		System.clearProperty(AotDetector.AOT_ENABLED);
+		
+		JpaRepositoryContributor contributor = new JpaRepositoryConfigExtension.JpaRepositoryRegistrationAotProcessor()
+				.contribute(new DummyAotRepositoryContext(context) {
+					@Override
+					public Set<Class<?>> getResolvedTypes() {
+						return Collections.singleton(Person.class);
+					}
+				}, ctx);
+
+		assertThat(contributor).isNull();
+	}
+
+	@Test // GH-3899
+	void repositoryProcessorShouldRespectExplicitRepositoryEnabledProperty() {
+
+		GenerationContext ctx = new DefaultGenerationContext(new ClassNameGenerator(ClassName.OBJECT),
+				new InMemoryGeneratedFiles());
+
+		GenericApplicationContext context = new GenericApplicationContext();
+		
+		System.setProperty(AotDetector.AOT_ENABLED, "true");
+
+		try {
+			MockPropertySource propertySource = new MockPropertySource()
```

Comment:
This could be handled by system properties as well as `GenericApplicationContext` creates `StandardEnvironment` that utilizes system properties and system environment for property sources. That would simplify the test arrangement.

Reply:
Good point. I changed to use `SetSystemProperty` for this!